From a693d49bff058fc20a448dc4e7d324ff0dc6597e Mon Sep 17 00:00:00 2001
From: Ignacio Casal Quinteiro <qignacio@amazon.com>
Date: Wed, 11 Sep 2024 11:52:11 +0200
Subject: [PATCH] websocket: process the frame as soon as we read data

Otherwise we can enter in a read loop because we were not
validating the data until the all the data was read.

Fixes #391

CVE: CVE-2024-52532
Upstream-Status: Backport [https://gitlab.gnome.org/GNOME/libsoup/-/commit/6adc0e3eb74c257ed4e2a23eb4b2774fdb0d67be#f1d67ca0386b145ea201cf88d27f72724d7c6715]

Upstream: https://git.openembedded.org/meta-openembedded/tree/meta-oe/recipes-support/libsoup/libsoup-2.4/CVE-2024-52532-1.patch

Signed-off-by: Changqing Li <changqing.li@windriver.com>
Signed-off-by: Titouan Christophe <titouan.christophe@mind.be>
---
 libsoup/soup-websocket-connection.c | 5 ++---
 1 file changed, 2 insertions(+), 3 deletions(-)

diff --git a/libsoup/soup-websocket-connection.c b/libsoup/soup-websocket-connection.c
index a4095e1..65c1492 100644
--- a/libsoup/soup-websocket-connection.c
+++ b/libsoup/soup-websocket-connection.c
@@ -1140,9 +1140,8 @@ soup_websocket_connection_read (SoupWebsocketConnection *self)
 		}
 
 		pv->incoming->len = len + count;
-	} while (count > 0);
-
-	process_incoming (self);
+		process_incoming (self);
+	} while (count > 0 && !pv->close_sent && !pv->io_closing);
 
 	if (end) {
 		if (!pv->close_sent || !pv->close_received) {
-- 
2.34.1

